File: Day13/VB.Net/Parser.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime.InteropServices

Namespace Day13VB
    Friend Class Parser
        Public Function Convert(ByVal matrix As Char()()) As Char(,)
            Dim w As Integer = matrix.Count()
            Dim h As Integer = matrix(0).Length
            Dim result = New Char(h - 1, w - 1) {}

            For i As Integer = 0 To w - 1

                For j As Integer = 0 To h - 1
                    result(j, i) = matrix(i)(j)
                Next
            Next

            Return result
        End Function

        Public Sub Load(ByRef cells As Char(,), ByRef carts As List(Of Cart))
            Dim file = System.IO.File.ReadAllLines("Input.txt").[Select](Function(line) line.ToCharArray()).ToArray()
            cells = Convert(file)
            carts = New List(Of Cart)()

            For y As Integer = 0 To cells.GetUpperBound(1)

                For x As Integer = 0 To cells.GetUpperBound(0)

                    Select Case cells(x, y)
                        Case "<"c
                            carts.Add(New Cart() With {
                                .X = x,
                                .Y = y,
                                .CurrentDirection = Direction.West,
                                .NextTurn = Turn.Left
                            })
                            cells(x, y) = "-"c
                        Case ">"c
                            carts.Add(New Cart() With {
                                .X = x,
                                .Y = y,
                                .CurrentDirection = Direction.East,
                                .NextTurn = Turn.Left
                            })
                            cells(x, y) = "-"c
                        Case "^"c
                            carts.Add(New Cart() With {
                                .X = x,
                                .Y = y,
                                .CurrentDirection = Direction.North,
                                .NextTurn = Turn.Left
                            })
                            cells(x, y) = "|"c
                        Case "v"c
                            carts.Add(New Cart() With {
                                .X = x,
                                .Y = y,
                                .CurrentDirection = Direction.South,
                                .NextTurn = Turn.Left
                            })
                            cells(x, y) = "|"c
                        Case Else
                    End Select
                Next
            Next
        End Sub
    End Class
End Namespace

File: Day13/Parser.cs
using System.Collections.Generic;
using System.Linq;

namespace Day13
{
    class Parser
    {
        public char[,] Convert(char[][] matrix)
        {
            int w = matrix.Count();
            int h = matrix[0].Length;

            var result = new char[h, w];

            for (int i = 0; i < w; i++)
            {
                for (int j = 0; j < h; j++)
                {
                    result[j, i] = matrix[i][j];
                }
            }

            return result;
        }

        public void Load(out char[,] cells, out List<Cart> carts)
        {
            var file = System.IO.File.ReadAllLines("Input.txt").Select(line => line.ToCharArray()).ToArray();
            cells = Convert(file);
            carts = new List<Cart>();
            for (int y = 0; y <= cells.GetUpperBound(1); y++)
            {
                for (int x = 0; x <= cells.GetUpperBound(0); x++)
                {

                    switch (cells[x, y])
                    {
                        case '<':
                            carts.Add(new Cart() { X = x, Y = y, CurrentDirection = Direction.West, NextTurn = Turn.Left });
                            cells[x, y] = '-';
                            break;

                        case '>':
                            carts.Add(new Cart() { X = x, Y = y, CurrentDirection = Direction.East, NextTurn = Turn.Left });
                            cells[x, y] = '-';
                            break;

                        case '^':
                            carts.Add(new Cart() { X = x, Y = y, CurrentDirection = Direction.North, NextTurn = Turn.Left });
                            cells[x, y] = '|';
                            break;

                        case 'v':
                            carts.Add(new Cart() { X = x, Y = y, CurrentDirection = Direction.South, NextTurn = Turn.Left });
                            cells[x, y] = '|';
                            break;

                        default:
                            break;
                    }
                }
            }
        }


    }
}
